Add 5.0 Actions Middleware change (#9943)
withastro/docs: src/content/docs/en/guides/actions.mdx

<!--...-->
```

### Advanced: Persist action results with a session

<p><Since v="5.0.0" /></p>

Action results are displayed as a POST submission. This means that the result will be reset to `undefined` when a user closes and revisits the page. The user will also see a "confirm form resubmission?" dialog if they attempt to refresh the page.

To customize this behavior, you can add middleware to handle the result of the action manually. You may choose the persist the action result using a cookie or session storage.

> Start by [creating a middleware file](https://docs.astro.build/en/guides/middleware/) and importing [the `getActionContext()` utility](/en/reference/modules/astro-actions/#getactioncontext) from `astro:actions`. This function returns an `action` object with information about the incoming action request, including the action handler and whether the action was called from an HTML form. `getActionContext()` also returns the `setActionResult()` and `serializeActionResult()` functions to programmatically set the value returned by `Astro.getActionResult()`:

Reviewer: nit: relative URL
```suggestion
Start by [creating a middleware file](/en/guides/middleware/) and importing [the `getActionContext()` utility](/en/reference/modules/astro-actions/#getactioncontext) from `astro:actions`. This function returns an `action` object with information about the incoming action request, including the action handler and whether the action was called from an HTML form. `getActionContext()` also returns the `setActionResult()` and `serializeActionResult()` functions to programmatically set the value returned by `Astro.getActionResult()`:
```

```ts title="src/middleware.ts" {2,5}
import { defineMiddleware } from 'astro:middleware';
import { getActionContext } from 'astro:actions';

export const onRequest = defineMiddleware(async (context, next) => {
  const { action, setActionResult, serializeActionResult } = getActionContext(context);
  if (action?.calledFrom === 'form') {
    const result = await action.handler();
    // ... handle the action result
    setActionResult(action.name, serializeActionResult(result));
  }
  return next();
});
```

A common practice to persist HTML form results is the [POST / Redirect / GET pattern](https://en.wikipedia.org/wiki/Post/Redirect/Get). This approach removes "confirm form resubmission?" dialog when a user refreshes by using a redirect, and gives you the option to persist action results throughout the user's session.

This example applies the POST / Redirect / GET pattern to all form submissions using session storage. Action results are written to a session store using [Netlify Blob](https://docs.netlify.com/blobs/overview/), and retrieved after a redirect using a session ID:

:::note

> This example requires the [Netlify server adapter](https://docs.astro.build/en/guides/integrations-guide/netlify/) to run. Feel free to adapt this example to your storage platform of choice.

Reviewer: nit: relative URL
```suggestion
This example requires the [Netlify server adapter](/en/guides/integrations-guide/netlify/) to run. Feel free to adapt this example to your storage platform of choice.
```

> This example requires the [Netlify server adapter](https://docs.astro.build/en/guides/integrations-guide/netlify/) to run. Feel free to adapt this example to your storage platform of choice.

Reviewer 2: Don't be fooled by this being outdated just because the link was updated! I suggest removing the note and working the context in normal paragraph text!

      actionName: action.name,
      actionResult: serializeActionResult(actionResult),
    });
  
    // Pass the session ID as a cookie
    // to be retrieved after redirecting to the page
    context.cookies.set("action-session-id", sessionId);
  
    // Redirect back to the previous page on error
    if (actionResult.error) {
      const referer = context.request.headers.get("Referer");
      if (!referer) {
        throw new Error(
          "Internal: Referer unexpectedly missing from Action POST request.",
        );
      }
      return context.redirect(referer);
    }
    // Redirect to the destination page on success
    return context.redirect(context.originPathname);
  }
  
  return next();
});
```

## Security when using actions

Actions are accessible as public endpoints based on the name of the action. For example, the action `blog.like()` will be accessible from `/_actions/blog.like`. This is useful for unit testing action results and debugging production errors. However, this means you **must** use same authorization checks that you would consider for API endpoints and on-demand rendered pages.

### Authorize users from an action handler

> The easiest way to authorize action requests is to add an authentication check to your action handler. You may reach for [an authentication library](https://docs.astro.build/en/guides/authentication/) to handle session management and user information. Actions expose the full `APIContext` object to access properties passed from middleware using `context.locals`. When a user is not authorized, you can raise an `ActionError` with the `UNAUTHORIZED` code:

Reviewer: nit: relative URL
```suggestion
The easiest way to authorize action requests is to add an authentication check to your action handler. You may reach for [an authentication library](/en/guides/authentication/) to handle session management and user information. Actions expose the full `APIContext` object to access properties passed from middleware using `context.locals`. When a user is not authorized, you can raise an `ActionError` with the `UNAUTHORIZED` code:
```

> The easiest way to authorize action requests is to add an authentication check to your action handler. You may reach for [an authentication library](https://docs.astro.build/en/guides/authentication/) to handle session management and user information. Actions expose the full `APIContext` object to access properties passed from middleware using `context.locals`. When a user is not authorized, you can raise an `ActionError` with the `UNAUTHORIZED` code:

Reviewer 2: ```suggestion
To authorize action requests, add an authentication check to your action handler. You may want to use [an authentication library](https://docs.astro.build/en/guides/authentication/) to handle session management and user information.

Actions expose the full `APIContext` object to access properties passed from middleware using `context.locals`. When a user is not authorized, you can raise an `ActionError` with the `UNAUTHORIZED` code:
```
Just felt like it wanted to be a new paragraph, I think? Also:
- don't tell people what is easy (and, it's ok to only tell them the happy path, even if there are more ways)
- I think something a little more bland than "reach for" is fitting for docs: "you may choose to use", "you may want to use", "you can use... or/instead of building your own authentication" Pick what you like, but less imagery, I think!

Reviewer 2: Same don't be fooled by outdated here!

```ts title="src/actions/index.ts" {6-8}
import { defineAction, ActionError } from 'astro:actions';

export const server = {
  getUserSettings: defineAction({
    handler: async (_input, context) => {
      if (!context.locals.user) {
        throw new ActionError({ code: 'UNAUTHORIZED' });
      }
      return { /* data on success */ };
    }
  })
}
```

### Gate actions from middleware

<p><Since v="5.0.0" /></p>

> Astro recommends authorizing user sessions from your action handler to respect permission levels and rate-limiting on a per-action basis. Still, you may want to gate requests to all actions (or a subset of actions) from middleware. 

Reviewer 2: ```suggestion
Astro recommends authorizing user sessions from your action handler to respect permission levels and rate-limiting on a per-action basis. However, you can also gate requests to all actions (or a subset of actions) from middleware.
```
"also" or "instead" works here. "However" is a more standard word than "still" (don't "reach for still" when you have a perfectly good however!)

Protip: most "you might want to"s can be tightened up without losing meaning, so we do try to do that whenever we can with reasonable effort.

> Use the `getActionContext()` function from your middleware to retrieve information about any inbound action requests. This including the action name and whether that action was called using a client-side RPC function (example: `actions.blog.like()`) or an HTML form.

Reviewer 2: ```suggestion
Use the `getActionContext()` function from your middleware to retrieve information about any inbound action requests. This includes the action name and whether that action was called using a client-side remote procedure call (RPC) function (e.g. `actions.blog.like()`) or an HTML form.
```
In general, we use the full name outside the parentheses and the abbreviation inside, but obviously we don't do this for EVERY thing (HTML, CSS) but we DO do it for MOST things. If you think this rises to the level of HTML/CSS, then we can consider not name checking it at all, or perhaps swapping and putting the full name in parentheses to show we have reader empathy. I'll need you to make the call here.

> This example rejects all action requests that do not have a valid session token. This is _not_ a substitute for authorization, though it does ensure that actions are only accessible when a session is present. If the check fails, a "Forbidden" response is returned:

Reviewer 2: This is honestly good, but I do worry that the one example we have in the "security" section is not good security?

I think what feels a tiny bit off is although you've clearly stated that this is not a substitute for authorization, the "though" feels a bit like "defending the practice" which maybe we don't need to do and maybe isn't the right angle in a section about security? So maybe something like:

```suggestion
The following example rejects all action requests that do not have a valid session token. If the check fails, a "Forbidden" response is returned. (Note: this method ensures that actions are only accessible when a session is present, but is _not_ a substitute for (proper? secure?) authorization.)
```

Meh, maybe splitting hairs. Offered as an example.

      actionName: action.name,
      actionResult: serializeActionResult(actionResult),
    });
  
    // Pass the session ID as a cookie
    // to be retrieved after redirecting to the page
    context.cookies.set("action-session-id", sessionId);
  
    // Redirect back to the previous page on error
    if (actionResult.error) {
      const referer = context.request.headers.get("Referer");
      if (!referer) {
        throw new Error(
          "Internal: Referer unexpectedly missing from Action POST request.",
        );
      }
      return context.redirect(referer);
    }
    // Redirect to the destination page on success
    return context.redirect(context.originPathname);
  }
  
  return next();
});
```

## Security when using actions

Actions are accessible as public endpoints based on the name of the action. For example, the action `blog.like()` will be accessible from `/_actions/blog.like`. This is useful for unit testing action results and debugging production errors. However, this means you **must** use same authorization checks that you would consider for API endpoints and on-demand rendered pages.

### Authorize users from an action handler

> To authorize action requests, add an authentication check to your action handler. You may want to use [an authentication library](https://docs.astro.build/en/guides/authentication/) to handle session management and user information.

Reviewer: I'm mostly a spectator here, sorry. But the changes are instructive! 👀 
As I was rereading, I noticed one remaining absolute URL:
```suggestion
To authorize action requests, add an authentication check to your action handler. You may want to use [an authentication library](/en/guides/authentication/) to handle session management and user information.
```
